cuML/DEVELOPER_GUIDE.md

```diff
@@ -0,0 +1,119 @@
+
+## Device and Host memory allocations
+To enable `libcuml.so` users to control how memory for temporary data is allocated device memory should only be allocated via the allocator provided:
+```cpp
+template<typename T>
+void foo(const ML::cumlHandle_impl& h, cudaStream_t stream, ... )
+{
+    T* temp_h = h.getDeviceAllocator()->allocate(n*sizeof(T), stream);
+    ...
+    h.getDeviceAllocator()->deallocate(temp_h, n*sizeof(T), stream);
+}
+```
+the same rule applies to larger amounts of host heap memory:
+```cpp
+template<typename T>
+void foo(const ML::cumlHandle_impl& h, cudaStream_t stream, ... )
+{
+    T* temp_h = h.getHostAllocator()->allocate(n*sizeof(T), stream);
+    ...
+    h.getHostAllocator()->deallocate(temp_h, n*sizeof(T), stream);
+}
+```
+Small host memory heap allocations, e.g. as internally done by STL containers, are fine, e.g. an `std::vector` managing only a handful of integers.
+Both the Host and the Device Allocators might allow asynchronous stream ordered allocation and deallocation. This can provide significant performance benefits so a stream always needs to be specified when allocating or deallocating (see [Asynchronous operations and stream ordering](# Asynchronous operations and stream ordering)).
+There are two simple container classes compatible with the allocator interface `MLCommon::device_buffer` available in `ml-prims/src/common/device_buffer.hpp` and `MLCommon::host_buffer` available in `ml-prims/src/common/host_buffer.hpp`. These allow to follow the [RAII idiom](https://en.wikipedia.org/wiki/Resource_acquisition_is_initialization) to avoid resources leaks and enable exception safe code. These containers also allow asynchronous allocation and deallocation using the `resize` and `release` member functions:
+```cpp
+template<typename T>
+void foo(const ML::cumlHandle_impl& h, ..., cudaStream_t stream )
+{
+    ...
+    MLCommon::device_buffer<T> temp( h.getDeviceAllocator(), stream, 0 )
+    
+    temp.resize(n, stream);
+    kernelA<<<grid, block, 0, stream>>>(..., temp.data(), ...);
+    kernelB<<<grid, block, 0, stream>>>(..., temp.data(), ...);
+    temp.release(stream);
+}
+```
+### Using thrust
+To ensure that thrust algorithms allocate temporary memory via the provided device memory allocator the `ML::thrustAllocatorAdapter` available in `allocatorAdapter.hpp` should be used with the `thrust::cuda::par` execution policy:
+```cpp
+void foo(const ML::cumlHandle_impl& h, ..., cudaStream_t stream )
+{
+    ML::thrustAllocatorAdapter alloc( h.getDeviceAllocator(), stream );
+    auto execution_policy = thrust::cuda::par(alloc).on(stream);
+    thrust::for_each(execution_policy, ... );
+}
+```
+The header `allocatorAdapter.hpp` also provides a helper function to create an execution policy:
+```cpp
+void foo(const ML::cumlHandle_impl& h, ... , cudaStream_t stream )
+{
+    auto execution_policy = ML::exec_policy(h.getDeviceAllocator(),stream);
```

Comment:
Regarding passing in `ML::cumHandle_impl&`: This is partly just my preference to be explicit, because the execution policy only needs the allocator and nothing else from handle. Besides that it avoids introducing a dependency of `ML::exec_policy` that is not really needed.
Regarding the one liner: I agree that this would be nicer. @jrhemstad said that this unfortunately requires thurst 1.9.4 or newer as the thrust execution policies with older thrust versions can't be initiated with temporaries. The current workaround with auto pointers also does not allow this because the auto pointer gets out of scope before the object is used. To my understanding another nice thing with thrust 1.9.4+ for this would be that one would not need to repeat the stream. We could provide a helper function that could be used like this:
```cpp
thrust::for_each(ML::exec_policy(alloc, stream), ... );
```